### PR TITLE
[FIX] base: allow duplicate IDs to be printed in a single report

### PR DESCRIPTION
**Pull request by ABE-Mark45**
### Current behavior before PR
The same item cannot be printed multiple times in a single report (as with the case with lot printing).

### Desired behavior after PR is merged
The same item can now be printed multiple times in a single report because resources IDs are now pushed into a `list` and not a `dict` with `ID` as the key.

task-3629043

image

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by ticodoo** [comment]
@ABE-Mark45 you need to clean up your commits first before marking this ready for review so that other teams don't spend time looking at it before it's ready. Your 1st + 3rd commits currently clash with each other. Logistics will also do the initial reviews before asking the other teams to review it.

File: addons/stock/report/report_lot_barcode.xml [hunk 1]
```diff
@@ -3,7 +3,7 @@
 <data>
 <template id="report_lot_label">
     <t t-call="web.basic_layout">
-        <t t-set='nRows' t-value='12'/>
+        <t t-set='nRows' t-value='7'/>
```
Comment (svs-odoo) [inline addons/stock/report/report_lot_barcode.xml:6]: Following my previous comment, we can change the size of the label itself to match the 4x12 format and change the label's size at the line 22:
```xml
<div t-att-style="'position:relative; width:43mm; height:19mm; border: 1px solid %s;' % (o.env.user.company_id.primary_color or 'black')">
```
I suggest you to use a height of 19mm ([like it was before](https://github.com/odoo/odoo/commit/2185c7c7303c2738c63ef454b061f85e009d29ee#diff-fbc7f34c633189f45e17080184e52bb403bf631e36d8713cb0b2d27260629a3bL21)) and I think the font size used for the product's name and the lot should be decreased accordingly

File: addons/product/static/src/scss/report_label_sheet.scss [hunk 2]
```diff
@@ -68,7 +68,7 @@
     }
 
     // generic 4x12 label w/ all same size font
-    div.o_label_4x12 {
+    div.o_label_4x12, div.o_label_4x7 {
```
Comment (svs-odoo) [inline addons/product/static/src/scss/report_label_sheet.scss:71]: Regarding my previous comment:
```suggestion
    .o_label_4x12 {
```
- Remove `o_label_4x7` since we want to keep the 4x12 format (and `o_label_4x7` wasn't used anymore anyway ?)
- Remove the `div` so the `o_label_4x12` rules aren't limited to div (I don't think it was wanted ? cc @ticodoo ) and the product's name and lot will be written smaller

[the remark this answers is not in the document]
Comment (ABE-Mark45) [reply addons/product/static/src/scss/report_label_sheet.scss:71]: Yes

File: odoo/addons/base/models/ir_actions_report.py [hunk 3]
```diff
@@ -103,13 +103,15 @@ def _split_table(tree, max_rows):
...
 
+
 class IrActionsReport(models.Model):
     _name = 'ir.actions.report'
     _description = 'Report Action'
     _inherit = 'ir.actions.actions'
     _table = 'ir_act_report_xml'
     _order = 'name'
     _allow_sudo_commands = False
+    StreamInfo = namedtuple('StreamInfo', ['id', 'data'])
```
[the remark this answers is not in the document]
Comment (ABE-Mark45) [reply odoo/addons/base/models/ir_actions_report.py:114]: To improve code readability. Using nested indexes was very hard to follow for me.

File: addons/product/static/src/scss/report_label_sheet.scss (hunk 2 above)
[the remark this answers is not in the document]
Comment (svs-odoo) [reply addons/product/static/src/scss/report_label_sheet.scss:71]: @ABE-Mark45 where then ? I didn't find it

**Comment by Whenrow** [comment]
@smetl Hello, could we have your review on the base/report part ? Thank you

File: odoo/addons/base/models/ir_actions_report.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (smetl) [reply odoo/addons/base/models/ir_actions_report.py:113]: @ABE-Mark45 It's not improving anything imo. Remove it.

**Review by smetl — COMMENTED** [review]
First of all, you can't change the current code like that in stable version.
Since you only want to print some barcodes... Why don't you just remove all res_ids? That way, all generated streams will be linked to `False` and just printed without any post-processing. Is it not what you want?

**Comment by ticodoo** [comment]
> First of all, you can't change the current code like that in stable version. Since you only want to print some barcodes... Why don't you just remove all res_ids? That way, all generated streams will be linked to `False` and just printed without any post-processing. Is it not what you want?

The issue is that in order to print out the same record in the same report multiple times we will have to write custom python code every time to create the `data` argument that contains all the info that's needed in the record instead of doing something like we currently do (`report_action([res_id] * qty`, used in at least 2 places not to mention potential future features) and we additionally will have to adapt the report template to handle the both the `res_ids` and `data` use cases (the report is used for both a "action > print" flow and a custom wizard flow that allows you to choose between printing the record once or a 'qty' amount). I.e. it's a huge pain the butt maintenance wise, makes our code much harder to understand, and seems like a regression of a feature that used to exist in the report engine.

So the things I'm wonder are:
- is there any value to using an `OrderedDict` vs a list?
- is there any value to using `set()` to compare the stream vs html? Do you expect there to ever be duplicate values?
- is this change at minimum worth doing in master?

**Comment by ticodoo** [comment]
@smetl  Thanks for the response, the logic is much clearer to me now. I think we will have to add an override hack like I previously did here:
https://github.com/odoo/odoo/blob/2f847c46292c409547cde402ae1dd38128eed4d2/addons/stock/models/ir_actions_report.py#L4-L16

and make sure that everywhere we need to print the same ID more than once will need to pass the `docids` via the `data` param in `report_action()` unless you have a better idea :crying_cat_face: 

Also @ABE-Mark45 if we do end up doing this hack as the fix (I already tested that it works in master), you need to target v16 (the earliest version where this feature is broken) and you'll have to apply it to different places where `report_lot_label` calls `.report_action()` and multiples the lot ids by the qty done (I think 2 places in v16 and 3 places in v17)

**Comment by ticodoo** [comment]
Should hopefully be fixed with https://github.com/odoo/odoo/pull/157710
